src/hotspot/cpu/x86/stubGenerator_x86_64_poly_mont.cpp

```diff
@@ -0,0 +1,429 @@
+
+  // Acc1 = 0
+  __ vpxorq(Acc1, Acc1, Acc1, Assembler::AVX_512bit);
+  for (int i = 0; i< 5; i++) {
+      // Acc2 = 0
+      __ vpxorq(Acc2, Acc2, Acc2, Assembler::AVX_512bit);
+
+      // B = replicate(bLimbs[i])
+      __ vpbroadcastq(B, Address(bLimbs, i*8), Assembler::AVX_512bit);
+
+      // Acc1 += A * B
+      __ evpmadd52luq(Acc1, A, B, Assembler::AVX_512bit);
+
+      // Acc2 += A *h B
+      __ evpmadd52huq(Acc2, A, B, Assembler::AVX_512bit);
+
+      // N = replicate(Acc1[0])
+      __ vpbroadcastq(N, Acc1, Assembler::AVX_512bit);
+
+      // Acc1 += M *  N
+      __ evpmadd52luq(Acc1, modulus, N, Assembler::AVX_512bit);
+
+      // Acc2 += M *h N
+      __ evpmadd52huq(Acc2, modulus, N, Assembler::AVX_512bit);
+
+      if (i == 4) break;
+
+      // Combine high/low partial sums Acc1 + Acc2
+
+      // carry = Acc1[0] >> 52
+      __ evpsrlq(carry, limb0, Acc1, 52, true, Assembler::AVX_512bit);
+
+      // Acc2[0] += carry
+      __ evpaddq(Acc2, limb0, carry, Acc2, true, Assembler::AVX_512bit);
+
+      // Acc1 = Acc1 shift one q element >>
+      __ evpermq(Acc1, allLimbs, shift1R, Acc1, false, Assembler::AVX_512bit);
+
+      // Acc1 = Acc1 + Acc2
+      __ vpaddq(Acc1, Acc1, Acc2, Assembler::AVX_512bit);
+  }
+
+  // Last Carry round: Combine high/low partial sums Acc1<high_bits> + Acc1 + Acc2
+  // carry = Acc1 >> 52
+  __ evpsrlq(carry, allLimbs, Acc1, 52, true, Assembler::AVX_512bit);
+
+  // Acc1 = Acc1 shift one q element >>
+  __ evpermq(Acc1, allLimbs, shift1R, Acc1, false, Assembler::AVX_512bit);
+
+  // Acc1  = mask52(Acc1)
+  __ evpandq(Acc1, Acc1, mask52, Assembler::AVX_512bit); // Clear top 12 bits
+
+  // Acc2 += carry
+  __ evpaddq(Acc2, allLimbs, carry, Acc2, true, Assembler::AVX_512bit);
+
+  // Acc1 = Acc1 + Acc2
+  __ vpaddq(Acc1, Acc1, Acc2, Assembler::AVX_512bit);
+
+  // output to rLimbs (1 + 4 limbs)
+  __ movq(Address(rLimbs, 0), Acc1);
+  __ evpermq(Acc1, k0, shift1R, Acc1, true, Assembler::AVX_512bit);
+  __ evmovdquq(Address(rLimbs, 8), k0, Acc1, true, Assembler::AVX_256bit);
+}
+
+address StubGenerator::generate_intpoly_montgomeryMult_P256() {
+  __ align(CodeEntryAlignment);
+  StubCodeMark mark(this, "StubRoutines", "intpoly_montgomeryMult_P256");
+  address start = __ pc();
+  __ enter();
+
+  // Save all 'SOE' registers
+  __ push(rbx);
+  #ifdef _WIN64
+  __ push(rsi);
+  __ push(rdi);
+  #endif
+  __ push(r12);
+  __ push(r13);
+  __ push(r14);
+  __ push(r15);
+
+  // Register Map
+  const Register aLimbs  = rdi;
+  const Register bLimbs  = rsi;
+  const Register rLimbs  = rdx;
+
+  // Normalize input
+  // pseudo-signature: void poly1305_processBlocks(byte[] input, int length, int[5] accumulator, int[5] R)
+  // a, b, r pointers point at first array element
+  // java headers bypassed in LibraryCallKit::inline_poly1305_processBlocks
+  #ifdef _WIN64
+  // c_rarg0 - rcx
+  // c_rarg1 - rdx
+  // c_rarg2 - r8
+  __ mov(aLimbs, c_rarg0);
+  __ mov(bLimbs, c_rarg1);
+  __ mov(rLimbs, c_rarg2);
```

Comment:
We could directly call montgomeryMultiply(c_rarg0, c_rarg1, c_rarg2) then these moves are not necessary.

Reply:
Gave them symbolic names and passed the gpr temp and parameter. vector register map still in the montgomeryMultiply function, but gprs explicitly passed in. 'close enough'?